src/administration/horizontal_scaling.md

```diff
@@ -0,0 +1,141 @@
+# Scaling Lemmy horizontally
+
+This is a collection of notes on scaling different Lemmy components horizontally. This is not meant as a step-by-step
+guide, rather as general tips and knowledge that might be useful to somebody who is already familiar with and horizontal scaling concepts and best
+practices. If you don't already know about concepts like load balancing,
+object storage, and reverse proxies, then this document will probably be hard (or impossible) to understand.
+
+**Note: scaling Lemmy this way is not necessary, and for small instances, the added overhead most likely
+outweighs any benefits.**
+
+## Why scale horizontally?
+
+- If one of your Lemmy servers dies for any reason, your instance can still remain operational
+- Having multiple Lemmy servers allows you to do rolling upgrades (no downtime needed to upgrade Lemmy!)
+- Horizontal scaling provides additional flexibility in how you upgrade your infrastructure
+- As of Lemmy version 0.18.2, Lemmy appears to be able to use the same resources more efficiently if they are split between multiple Lemmy processes
+
+## Breakdown of Lemmy components
+
+### Lemmy-ui
+
+Lemmy-ui is the main frontend for Lemmy. It consists of an expressjs based server-side process (necessary for SSR) and
+client code which run in the browser. It does not use a lot of resources and will happily run on quite low powered
+servers.
+
+#### Rolling upgrades
+
+In general, there is nothing preventing you from running multiple load balanced Lemmy-ui servers in parallel, but
+without some custom configuration, **rolling upgrades will break Lemmy for your users!** This is because by default,
+Lemmy-ui will serve static files through the express process.
+
+Consider the scenario where you have 2 Lemmy-ui servers, one is running on version 1, and the other has just come online
+running version 2. A user opening your Lemmy in their browser might have their front page html request routed to version 1.
+This html will contain a reference to many other static files which are specific to version 1. Most likely, some of these
+requests will get routed to the server running version 2. **The files will not exist on this server, the server will
```

Comment:
I would always recommend turning on ip hash load leveling. that should prevent most of this?

Reply:
Seems like the problem should be prevented by [this change](https://github.com/LemmyNet/lemmy-ui/pull/1684). Or is it not working?

Reply:
>@phiresky I would always recommend turning on ip hash load leveling. that should prevent most of this?

It's similar to sticky sessions, I'll add it to the list 

> @nutomic Seems like the problem should be prevented by this change. Or is it not working?

That change is working without problems, but it doesn't help here - different lemmy-ui versions will have different hashes. So out of the box, if you request `/static/<hash_from_version_1>` from a lemmy-ui running version 2, you will get a 404 response.

Reply:
I added the note about ip hash based load balancing